### PR TITLE
Fixes #4003 : flag EI_EXPOSE_REP in package-private classes overriding public super-type methods

### LINKED ISSUE #4003: Public interface and shadow copies lead to false negatives in EI_EXPOSE_REP(2)
**Describe the bug**
I find some false negatives in the `FindReturnRef` detector responsible for the `EI_EXPOSE_REP` and `EI_EXPOSE_REP2` vulnerability rules. 
The rule ignores package-private classes. If a package-private class implements a `public interface` and exposes its internal mutable state through overridden public methods, SpotBugs fails to flag it.

And, I also find that the rule seems to assume that invoking `.clone()` on an array mitigates representation exposure. However, for multi-dimensional arrays (e.g., `byte[][]`) or arrays of mutable objects (e.g, HashTable), `.clone()` may only perform a *shallow copy*. The inner mutable objects are still exposed by reference, but now the rule downgrades or the finding to `LOW_PRIORITY` (Priority 3). I am not sure if this is expected.

**To Reproduce**
```java
// CASE 1: Scope Filtering False Negative
// SpotBugs ignores package-private classes, missing exposures
// that escape through public interfaces to external callers.
public interface Provider { 
    byte[] getData(); 
}
class HiddenProvider implements Provider {
    private byte[] data = {1, 2, 3};
    @Override
    public byte[] getData() {
        // FALSE NEGATIVE: Returns direct reference.
        // Missed entirely because the declaring class is not 'public'.
        return data; 
    }
}
// CASE 2 & 3: Shallow Clone 
public class VulnerableManager {
    private byte[][] matrix;
    private Hashtable<String, byte[]> cache;
    // CASE 2: Complex Array Shallow Clone
    public byte[][] getMatrix() {
        // Downgraded to Priority 3 / Low:
        return matrix.clone(); 
    }
    // CASE 3: Mutable Collection Shallow Clone
    public Hashtable<String, byte[]> getCache() {
        // FALSE NEGATIVE:
        // Both keys in the Hashtable are shallow copies.
        return (Hashtable<String, byte[]>) cache.clone(); 
    }
}
```

**Expected behavior**

**Environment**
- Java Compiler: javac 21.0.10
- Java version: JDK 21
- SpotBugs Version: 4.9.8
- How do you use SpotBugs: CLI

### PR DESCRIPTION
**Pull request by mirkoalicastro**
`FindReturnRef` used to skip any non-public class (`getThisClass().isPublic() && obj.isPublic()`), so that a package-private class that implements a public interface and exposes a private mutable field through an overridden public method was never flagged, even if the exposure is reachable from outside the package via the public super-type.

With these changes, the detector still skips non-public methods, but a package-private declaring class is analysed when the public method overrides one declared in an externally-visible super-type. Static methods are excluded from the new path since they aren't subject to virtual dispatch through a super-type.

Added a unit test reproducing the issue #4003 . Without the changes, the unit test would fail and the violation would not be detected.

- [x] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

**Review by ThrawnCA — APPROVED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java [hunk 1]
```diff
@@ -314,6 +320,23 @@ public void sawOpcode(int seen) {
...
+        XMethod xMethod = XFactory.createXMethod(getThisClass(), method);
+        for (XMethod superMethod : Hierarchy2.findSuperMethods(xMethod)) {
+            try {
+                if (superMethod.getClassDescriptor().getXClass().isPublic()) {
+                    return true;
+                }
+            } catch (CheckedAnalysisException e) {
+                AnalysisContext.logError("Error checking visibility of " + superMethod.getClassDescriptor(), e);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (JuditKnoll) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java:332]: This is by design like this.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java [hunk 2]
```diff
@@ -152,7 +154,11 @@ private void collectData(JavaClass javaClass, Method m) {
 
     @Override
     public void visit(Method obj) {
-        check = getThisClass().isPublic() && obj.isPublic();
+        if (!obj.isPublic()) {
+            check = false;
+            return;
+        }
+        check = getThisClass().isPublic() || overridesMethodFromPublicSupertype(obj);
```
Comment (JuditKnoll) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java]: If I'm not mistaken, this could be simplified with assigning a value only once to check like this:
```
check = obj.isPublic() || getThisClass().isPublic() || overridesMethodFromPublicSupertype(obj);
```

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java [hunk 3]
```diff
@@ -286,6 +286,15 @@ void testUnmodifiableClass() {
         assertBugTypeCount("MS_EXPOSE_REP", 0);
     }
 
+    @Test
+    void testPackagePrivateImplOfPublicInterface() {
+        performAnalysis("exposemutable/PackagePrivateImplProvider.class",
+                "exposemutable/PackagePrivateHiddenProvider.class");
+
+        assertBugTypeCount("EI_EXPOSE_REP", 1);
```
Comment (JuditKnoll) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java:296]: Can you please also assert the expected number of other similar bugtypes (which are checked in the `assertNoExposeBug()` function) as well, even if 0 hit is expected?

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java [hunk 4]
```diff
@@ -286,6 +286,15 @@ void testUnmodifiableClass() {
         assertBugTypeCount("MS_EXPOSE_REP", 0);
     }
 
+    @Test
+    void testPackagePrivateImplOfPublicInterface() {
```
Comment (JuditKnoll) [inline spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java:290]: Can you please add a few more tests? Both positive (there is a bug expected) and negative tests would be nice.

**Review by JuditKnoll — CHANGES_REQUESTED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java (hunk 2 above)
Comment (mirkoalicastro) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindReturnRef.java]: I think that using OR between all the three is too broad because it will check non-public methods in public classes. 
I've simplified it with:
```
check = obj.isPublic() && (getThisClass().isPublic() || overridesMethodFromPublicSupertype(obj));
```

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java (hunk 4 above)
Comment (mirkoalicastro) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java:290]: Added, thanks

File: spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java (hunk 3 above)
Comment (mirkoalicastro) [reply spotbugs-tests/src/test/java/edu/umd/cs/findbugs/detect/FindReturnRefTest.java:296]: Done, thanks

**Review by JuditKnoll — APPROVED** [review]

**Review by hazendaz — APPROVED** [review]
